### PR TITLE
[Client] Use a single RPC to fetch ClientObjectRefs passed in a list

### PR DESCRIPTION
**Pull request by mwtian**
## Why are these changes needed?
Currently when handling `ray.get()` on a list of `ClientObjectRef`, Ray client issues separate RPCs per `ClientObjectRef` and waiting on each to return. This magnifies the client and server round trip latency by the number of items in the list, and prevent parallel fetching of objects.

A benchmark is added to measure scheduling 1000 tasks and getting back the array.
- Before the change: `client: tasks and get batch per second 0.72 +- 0.0`.
- After the change: `client: tasks and get batch per second 1.11 +- 0.01`.

Also add a benchmark for putting large (1GB) objects via Ray client.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/protobuf/ray_client.proto [hunk 1]
```diff
@@ -114,7 +114,7 @@ message PutResponse {
 // Requests data from the server.
 message GetRequest {
   // The reference ID for the requested object data
-  bytes id = 1;
```
Comment (WangTaoTheTonic) [inline src/ray/protobuf/ray_client.proto:117]: Is it possible to change this field name to plural?

**Review by WangTaoTheTonic — APPROVED** [review]
Nice improment！LGTM with just a tiny comment.

File: src/ray/protobuf/ray_client.proto (hunk 1 above)
Comment (mwtian) [reply src/ray/protobuf/ray_client.proto:117]: Good suggestion, updated.

Comment (ijrsvt) [reply src/ray/protobuf/ray_client.proto:117]: Wait, @mwtian I strongly prefer following the convention of not modifying existing proto fields. Can we mark `id` as deprecated and add a new field `ids`

Comment (ijrsvt) [reply src/ray/protobuf/ray_client.proto:117]: Also make sure to update the PROTO_VERSION const!

File: python/ray/util/client/worker.py [hunk 2]
```diff
@@ -190,51 +189,26 @@ def connection_info(self):
...
-            deadline = time.monotonic() + timeout
-        out = []
-        for obj_ref in to_get:
-            res = None
-            # Implement non-blocking get with a short-polling loop. This allows
-            # cancellation of gets via Ctrl-C, since we never block for long.
-            while True:
-                try:
-                    if deadline:
-                        op_timeout = min(
```
Comment (ijrsvt) [inline python/ray/util/client/worker.py:220]: Can we keep the `MAX_BLOCKING_OPERATION`. This is still relevant because the server will block for the duration of the Ray Get!

**Review by ijrsvt — COMMENTED** [review]
Can we add unit tests for this?

File: python/ray/ray_perf.py [hunk 3]
```diff
@@ -62,13 +62,6 @@ def small_value():
     return b"ok"
```
Comment (ijrsvt) [inline python/ray/ray_perf.py:64]: Out of Curiosity, what does this change do?

File: python/ray/util/client/worker.py (hunk 2 above)
Comment (mwtian) [reply python/ray/util/client/worker.py:220]: Misunderstood how threading.Condition() works. Reverted this part. Will convert this to asyncio in a subsequent pr.

File: python/ray/tests/test_client.py [hunk 4]
```diff
@@ -386,6 +398,7 @@ def test_log(level, msg):
         assert ray.get(x) == "Foo"
         time.sleep(1)
         logs_with_id = [msg for msg in log_msgs if msg.find(x.id.hex()) >= 0]
+        print(logs_with_id)
```
Comment (ijrsvt) [inline python/ray/tests/test_client.py:401]: Is this intended?

File: src/ray/protobuf/ray_client.proto (hunk 1 above)
Comment (ijrsvt) [reply src/ray/protobuf/ray_client.proto:117]: This should be fine given that we updated the protocol. In general I usually prefer something like:
```
bytes id = 1 [deprecated = true];
float timeout = 2;
bool asynchronous = 3;
repeated bytes ids = 4;

```

File: python/ray/tests/test_client.py [hunk 5]
```diff
@@ -284,6 +284,18 @@ def f():
...
 
+@pytest.mark.skipif(sys.platform == "win32", reason="Failing on Windows.")
+def test_function_and_get_batch(ray_start_regular_shared):
+    with ray_start_client_server() as ray:
+
+        @ray.remote
+        def f():
+            return "OK"
+
+        assert ray.get(
```
Comment (ijrsvt) [inline python/ray/tests/test_client.py:295]: Hmm, could we break the abstraction barrier and look at the `ray.worker.dataclient.req_id` and see if it only gets incremented by one?

Comment (mwtian) [reply python/ray/tests/test_client.py]: Done.

File: python/ray/tests/test_client.py (hunk 4 above)
Comment (mwtian) [reply python/ray/tests/test_client.py]: No, changed to print under assertion failure.

**Comment by mwtian** [comment]
Debugging the failure of `//python/ray/tests:test_basic_3_client_mode::test_background_tasks_with_max_calls`. There seems to be a deadlock when >= 8 tasks are started [here](https://github.com/ray-project/ray/blob/55a90e670a5c71eea52876f8060dde0c7562c2d3/python/ray/tests/test_basic_3.py#L156).

**Review by ericl — APPROVED** [review]

**Comment by AmeerHajAli** [comment]
@mwtian , any updates on this?

**Comment by mwtian** [comment]
> @mwtian , any updates on this?

Starting to debug the remaining test failure.

File: python/ray/_private/ray_client_microbenchmark.py [hunk 6]
```diff
@@ -32,7 +46,16 @@ def do_put_small():
     def put_multi_small():
         ray.get([do_put_small.remote() for _ in range(10)])
 
-    results += timeit("client: remote put calls", put_multi_small, 1000)
+    results += timeit("client: tasks and put batch", put_multi_small, 1000)
+
+
+def benchmark_put_large(ray, results):
+    arr = np.zeros(100 * 1024 * 1024, dtype=np.int64)
```
[the remark this answers is not in the document]
Comment (mwtian) [reply python/ray/_private/ray_client_microbenchmark.py:53]: Yes, this is just copying data.

File: python/ray/tests/test_basic_3.py [hunk 7]
```diff
@@ -160,7 +160,7 @@ def f():
         pid, g_id = nested.pop(0)
         ray.get(g_id)
         del g_id
-        wait_for_pid_to_exit(pid)
+        wait_for_pid_to_exit(pid, timeout=900)
```
Comment (AmeerHajAli) [inline python/ray/tests/test_basic_3.py:163]: Why is this necessary? It looks like it is hiding some issue.

Comment (mwtian) [reply python/ray/tests/test_basic_3.py:163]: This is for debugging. Feel free to take a look later after @author-action-required is removed.

Comment (AmeerHajAli) [reply python/ray/tests/test_basic_3.py:163]: Oh, my bad. I got a notification "@mwtian mwtian requested a review from AmeerHajAli as a code owner 4 days ago" so I thought you forgot to remove it.

Comment (mwtian) [reply python/ray/tests/test_basic_3.py]: Ready now.

**Comment by mwtian** [comment]
Windows build seems to be broken on `master`: https://github.com/ray-project/ray/runs/3459113393

File: python/ray/tests/test_basic_3.py [hunk 8]
```diff
@@ -156,10 +157,17 @@ def f():
...
         pid, g_id = nested.pop(0)
-        ray.get(g_id)
-        del g_id
+        pids.append(pid)
+        assert ray.get(g_id) == 0
+        g_id = None
+
+    nested = None
+    # Force dereference all objects, so workers can exit.
+    gc.collect()
```
Comment (ckw017) [inline python/ray/tests/test_basic_3.py:169]: Oh wow, great catch!

**Review by ckw017 — APPROVED** [review]
LGTM, though noticed that the ray_client.proto doesn't have the `bytes id = 1 [deprecated = true];` suggestion

File: python/ray/tests/test_basic_3.py [hunk 9]
```diff
@@ -156,10 +157,17 @@ def f():
...
     while nested:
         pid, g_id = nested.pop(0)
-        ray.get(g_id)
-        del g_id
+        pids.append(pid)
+        assert ray.get(g_id) == 0
+        g_id = None
+
+    nested = None
+    # Force dereference all objects, so workers can exit.
```
Comment (ijrsvt) [inline python/ray/tests/test_basic_3.py:168]: What is the main reason for this change?

**Review by ijrsvt — APPROVED** [review]
One comment, but LGTM overall!

File: python/ray/tests/test_basic_3.py (hunk 9 above)
Comment (mwtian) [reply python/ray/tests/test_basic_3.py]: Simplified the change to just running GC. Basically without explicit GC, there would be leftover ObjectRefs in workers, preventing workers from exiting. A possible cause is that using the list of objects refs for `ray.get()`, compared to using ObjectRef individually, has different behavior under Python GC which affects Ray ref counting. But I don't have more insights into the GC behavior yet.

**Comment by mwtian** [comment]
> LGTM, though noticed that the ray_client.proto doesn't have the `bytes id = 1 [deprecated = true];` suggestion

Client version number is already updated so there should be no compatibility issue. Since both Ian and you raised this, I made the change to use a different tag.
